### PR TITLE
Replace `Vec::new` with `with_capacity` if possible

### LINKED ISSUE #436: Audit all `Vec::new` in the code and change to `Vec::with_capacity` if possible

### PR DESCRIPTION
**Pull request by duguorong009**
## Motivation
Resolve #436 

## Proposal
Audit all `Vec::new` & replace them with `Vec::with_capacity` if possible

## Test Plan
CI

### CODE REVIEW

File: linera-service/src/linera/main.rs [hunk 1]
```diff
@@ -1399,7 +1399,7 @@ async fn run(options: &ClientOptions) -> anyhow::Result<()> {
             )?;
             let mut genesis_config_ref = Persist::mutate(&mut genesis_config);
             let mut rng = Box::<dyn CryptoRng>::from(*testing_prng_seed);
-            let mut chains = vec![];
+            let mut chains = Vec::with_capacity(*num_other_initial_chains as usize);
```
Comment (afck) [inline linera-service/src/linera/main.rs:1402]: Let's use `usize::try_from` and fail if it's too large.

File: linera-storage/src/db_storage.rs [hunk 2]
```diff
@@ -474,13 +474,14 @@ where
...
-        let mut keys = Vec::new();
+        let mut keys = Vec::with_capacity(blob_ids.len());
         for blob_id in blob_ids.clone() {
             let key = bcs::to_bytes(&BaseKey::Blob(blob_id))?;
             keys.push(key);
         }
         let results = self.client.client.contains_keys(keys).await?;
-        let mut missing_blobs = Vec::new();
+        let count_false = results.iter().filter(|&&result| !result).count();
+        let mut missing_blobs = Vec::with_capacity(count_false);
```
Comment (afck) [inline linera-storage/src/db_storage.rs:484]: Let's not use it here.

File: linera-storage/src/db_storage.rs [hunk 3]
```diff
@@ -474,13 +474,14 @@ where
     }
 
     async fn missing_blobs(&self, blob_ids: Vec<BlobId>) -> Result<Vec<BlobId>, ViewError> {
-        let mut keys = Vec::new();
+        let mut keys = Vec::with_capacity(blob_ids.len());
         for blob_id in blob_ids.clone() {
```
Comment (afck) [inline linera-storage/src/db_storage.rs:478]: (Not your fault, but: No need to allocate a new vector.)

```suggestion
        for blob_id in blob_ids.iter().cloned() {
```

File: linera-storage/src/db_storage.rs [hunk 4]
```diff
@@ -575,7 +576,7 @@ where
         limit: u32,
     ) -> Result<Vec<HashedCertificateValue>, ViewError> {
         let mut hash = Some(from);
-        let mut values = Vec::new();
+        let mut values = Vec::with_capacity(limit as usize);
```
Comment (afck) [inline linera-storage/src/db_storage.rs:579]: This loop has `break`s in it, and often won't actually yield that many elements.

**Comment by afck** [comment]
Thank you for helping us explore this idea!

We had a discussion and decided to not do this for now, because it makes it easy to allocate too much in the end: E.g. one can overlook or insert `break` statements of make the `push` conditional without updating the `with_capacity`.

In some places we could use `Iterator::collect` instead: For `ExactSizeIterator`s that automatically allocates the right amount. (But only where that doesn't make the code less readable.)

**Comment by duguorong009** [comment]
> Thank you for helping us explore this idea!
> 
> We had a discussion and decided to not do this for now, because it makes it easy to allocate too much in the end: E.g. one can overlook or insert `break` statements of make the `push` conditional without updating the `with_capacity`.
> 
> In some places we could use `Iterator::collect` instead: For `ExactSizeIterator`s that automatically allocates the right amount. (But only where that doesn't make the code less readable.)

Thanks for quick review!
Yes, I agree that it could be better to allow `Vec` reallocate than pre-allocation.

**Pull request closed without merging**